src/os/bluestore/BlueStore.h

```diff
@@ -454,25 +465,25 @@ class BlueStore : public ObjectStore,
       _add_buffer(cache, b, (flags & Buffer::FLAG_NOCACHE) ? 0 : 1, nullptr);
       cache->_trim();
     }
-    void _finish_write(BufferCacheShard* cache, uint64_t seq);
-    void did_read(BufferCacheShard* cache, uint32_t offset, ceph::buffer::list& bl) {
+    void _finish_write(BufferCacheShard* cache, uint64_t offset, uint64_t seq);
+    void did_read(BufferCacheShard* cache, uint64_t offset, uint64_t length, ceph::buffer::list&& bl) {
       std::lock_guard l(cache->lock);
-      Buffer *b = new Buffer(this, Buffer::STATE_CLEAN, 0, offset, bl);
-      b->cache_private = _discard(cache, offset, bl.length());
+      Buffer *b = new Buffer(this, Buffer::STATE_CLEAN, 0, offset, std::move(bl));
```

Comment:
Interesting. Has this improvement been possible before?

Reply:
I don't think so. This buffer is being moved because we have to substr the previously padded data, therefore this case can just take ownership of the bl passed.
The other case has to copy the padded buffer so that it's stored in buffer.

I believe we could make improvements by following a close to 0 copy of bufferlist data by just reusing buffers created while doing a write.

For example in the context of main branch: we do `_do_alloc_write` and we perform two sequential operations like this:

```c++
    _buffer_cache_write(txc, wi.b, o, wi.logical_offset, wi.bl,
                        wctx->buffered ? 0 : Buffer::FLAG_NOCACHE);
    wi.b->get_blob().map_bl(
	  b_off, *l,
	  [&](uint64_t offset, bufferlist& t) {
          bdev->aio_write(offset, t, &txc->ioc, false);
   });
``` 

If you think about it here we have 1 copy of `wi.bl` inside `_buffer_cache_write` when doing `new Buffer`. It is completely unnecessary, we could just do something like:

```
    Buffer* buffer =_buffer_cache_write(txc, wi.b, o, wi.logical_offset, wi.bl,
                        wctx->buffered ? 0 : Buffer::FLAG_NOCACHE);
    wi.b->get_blob().map_bl(
	  b_off, *buffer->data, // <<<<<<<<<<<<<<<<<< reference of data
	  [&](uint64_t offset, bufferlist& t) {
          bdev->aio_write(offset, t, &txc->ioc, false);
   });
```